services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,517 @@
+     *         tenant.</li>
+     *         <li><em>412 Precondition Failed</em> if the version exists but mismatch the current tenants version.</li>
+     *         </ul>
+     */
+    @Override
+    public Future<OperationResult<Void>> updateTenant(final String tenantId, final Tenant tenantObj,
+            final Optional<String> resourceVersion, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantId)
+                .compose(ok -> findTenant(tenantId))
+                .compose(tenantDto -> checkConflictingCATenants(Optional.of(tenantDto),
+                        tenantObj.getTrustedCertificateAuthoritySubjectDNs(), span))
+                .compose(tenantDto -> processUpdateTenant(tenantId, tenantObj, resourceVersion, span))
+                .recover(error -> Future.succeededFuture(MongoDbDeviceRegistryUtils.mapErrorToResult(error, span)));
+    }
+
+    private Future<OperationResult<Void>> processUpdateTenant(final String tenantId, final Tenant newTenant,
+            final Optional<String> resourceVersion, final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(newTenant);
+        Objects.requireNonNull(resourceVersion);
+        Objects.requireNonNull(span);
+
+        final JsonObject updateTenantQuery = resourceVersion
+                .map(version -> new MongoDbDocumentBuilder().withVersion(version))
+                .orElse(new MongoDbDocumentBuilder())
+                .withTenantId(tenantId)
+                .document();
+
+        final Promise<MongoClientUpdateResult> updateTenantPromise = Promise.promise();
+        final TenantDto newTenantDto = new TenantDto(tenantId, newTenant,
+                new Versioned<>(newTenant).getVersion());
+        mongoClient.updateCollection(config.getCollectionName(), updateTenantQuery,
+                new JsonObject().put("$set", JsonObject.mapFrom(newTenantDto)), updateTenantPromise);
+
+        return updateTenantPromise.future()
+                .compose(updateResult -> {
+                    if (updateResult.getDocMatched() == 0) {
+                        return MongoDbDeviceRegistryUtils.checkForVersionMismatchAndFail(tenantId, resourceVersion,
+                                findTenant(tenantId));
+                    } else {
+                        span.log(String.format("successfully updated tenant [%s]", tenantId));
+                        return Future.succeededFuture(OperationResult.ok(
+                                HttpURLConnection.HTTP_NO_CONTENT,
+                                null,
+                                Optional.empty(),
+                                Optional.of(newTenantDto.getVersion())));
```

Comment:
Here we assume based on the number of records that the update has been successful and hence we send the `newTenantDto.getVersion()`. IMHO it would have been even better, if we are able to get the version of that  document after update. For that `findOneAndReplaceWithOptions(...)` would be more suited. Device registration service uses it too.

Reply:
done